### PR TITLE
ESQL - generate docs for snapshot functions

### PR DESCRIPTION
**Pull request by carlosdelest**
After conversations in [this PR](https://github.com/elastic/elasticsearch/pull/112590#discussion_r1760569431), we'll be adding docs generation to functions available on snapshot builds.

This PR adds that capability, and includes the generated docs for the `Categorize` function

### CODE REVIEW

**Comment by ioanatia** [comment]
cc @jan-elastic since the `categorize` function was added in https://github.com/elastic/elasticsearch/pull/112860

there is also the `rate` function that is available in snapshots only - @dnhatn is it necessary we generate Kibana docs for `rate` as well?

**Comment by carlosdelest** [comment]
> there is also the rate function that is available in snapshots only - @dnhatn is it necessary we generate Kibana docs for rate as well?

Rate still has no `AbstractFunctionTestCase` implementation - no docs can be generated from it yet.

**Review by fang-xing-esql — APPROVED** [review]
LGTM, thank you @carlosdelest !

File: docs/reference/esql/functions/layout/categorize.asciidoc
[the remark this answers is not in the document]
Comment (nik9000) [reply docs/reference/esql/functions/layout/categorize.asciidoc:1]: Making the files without plugging them in shouldn't hurt anything if that's what we want to do. I *think* we're ok doing that. I am. We should just make sure folks know having the generated asciidoc doesn't mean it's there - you have to look for the inclusion. For qstr I believe the plan is to include it only in snapshot docs. INLINESTATS works that way:

```
ifeval::["{release-state}"=="unreleased"]
* experimental:[] <<esql-inlinestats-by>>
endif::[]

...

ifeval::["{release-state}"=="unreleased"]
include::processing-commands/inlinestats.asciidoc[]
endif::[]

```

[the remark this answers is not in the document]
Comment (ivancea) [reply docs/reference/esql/functions/layout/categorize.asciidoc:1]: > Making the files without plugging them in shouldn't hurt anything

Indeed, but I feel like adding those without actually including them leaves things half-done. And it's harder later to know which ones are missing.

Btw, didn't know we had those conditionals! I'd include it with them, and that's all. About making sure it's done, it feels a bit more complex, and I can't think of a good way that doesn't pass through autogenerating those files too 👀 (Which feels like the eventual solution...)

[the remark this answers is not in the document]
Comment (carlosdelest) [reply docs/reference/esql/functions/layout/categorize.asciidoc:1]: That is correct - for now, we're not including the docs as these functions are not production ready yet.

Including functions docs should be ok if they are not linked to the global docs - they can exist but not be connected until we do so for production readiness.

**Pull request merged**